[while_loop] data-dependent op in body_fn

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141393
* #140106
* __->__ #142031
* #142162

The idea is the parent hop's fake tensor mode should ignore the newly allocated unbacked symints in subgraph because the bindings of unbacked symbols in the subgraph should already be done when we trace the subgraph. E.g. if there's an operator in subgraph that produces unbacked symints, the track_tensor_tree logic for that operator will take care of it.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

ydwu4 committed:
[while_loop] data-dependent op in body_fn

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] data-dependent op in body_fn"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] data-dependent op in body_fn"


The idea is  that the bindings of unbacked symbols within the subgraph should be done when we trace the subgraph. Specifically, if there's an operator in subgraph that produces unbacked symints, the track_tensor_tree logic for that operator will take care of it.

If the unbacked symint produced by subgraph is part of hop's output (e.g. the body_fn's output contains the unbacked symint), parent graph should create new unbacked symint for the output symint and let the track_tensor_tree logic of parent graph to take care of the new unbacked symints.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] data-dependent op in body_fn"


The idea is  that the bindings of unbacked symbols within the subgraph should be done when we trace the subgraph. Specifically, if there's an operator in subgraph that produces unbacked symints, the track_tensor_tree logic for that operator will take care of it.

If the unbacked symint produced by subgraph is part of hop's output (e.g. the body_fn's output contains the unbacked symint), parent graph should create new unbacked symint for the output symint and let the track_tensor_tree logic of parent graph to take care of the new unbacked symints.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] data-dependent op in body_fn"


The idea is  that the bindings of unbacked symbols within the subgraph should be done when we trace the subgraph. Specifically, if there's an operator in subgraph that produces unbacked symints, the track_tensor_tree logic for that operator will take care of it.

If the unbacked symint produced by subgraph is part of hop's output (e.g. the body_fn's output contains the unbacked symint), parent graph should create new unbacked symint for the output symint and let the track_tensor_tree logic of parent graph to take care of the new unbacked symints.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] data-dependent op in body_fn"


The idea is  that the bindings of unbacked symbols within the subgraph should be done when we trace the subgraph. Specifically, if there's an operator in subgraph that produces unbacked symints, the track_tensor_tree logic for that operator will take care of it.

If the unbacked symint produced by subgraph is part of hop's output (e.g. the body_fn's output contains the unbacked symint), parent graph should create new unbacked symint for the output symint and let the track_tensor_tree logic of parent graph to take care of the new unbacked symints.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] data-dependent op in body_fn"


The idea is  that the bindings of unbacked symbols within the subgraph should be done when we trace the subgraph. Specifically, if there's an operator in subgraph that produces unbacked symints, the track_tensor_tree logic for that operator will take care of it.

If the unbacked symint produced by subgraph is part of hop's output (e.g. the body_fn's output contains the unbacked symint), parent graph should create new unbacked symint for the output symint and let the track_tensor_tree logic of parent graph to take care of the new unbacked symints.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] data-dependent op in body_fn"


The idea is  that the bindings of unbacked symbols within the subgraph should be done when we trace the subgraph. Specifically, if there's an operator in subgraph that produces unbacked symints, the track_tensor_tree logic for that operator will take care of it.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/while_loop.py:
```diff
@@ -250,7 +248,41 @@ def while_loop_fake_tensor_mode(
+        with mode.shape_env.ignore_fresh_unbacked_symbols():
+            # body_fn return output with the same pytree and tensor meta data as carried_inputs
+            # so we could just return the output after one iteration.
+            return body_fn(*carried_inputs, *additional_inputs)
```
discussed offline the metadata check should be in the fake kernel here

ydwu4 replied:
Done! We cannot share the check (we could share the utils though) with dynamo as dynamo doesn't do the fake propagation of while_loop so we have to have two checks: one in fake tensor mode, the other in dynamo..

ydwu4 committed:
Update on "[while_loop] data-dependent op in body_fn"


The idea is the parent hop's fake tensor mode should ignore the newly allocated unbacked symints in subgraph because the bindings of unbacked symbols in the subgraph should already be done when we trace the subgraph. E.g. if there's an operator in subgraph that produces unbacked symints, the track_tensor_tree logic for that operator will take care of it.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] data-dependent op in body_fn"


The idea is the parent hop's fake tensor mode should ignore the newly allocated unbacked symints in subgraph because the bindings of unbacked symbols in the subgraph should already be done when we trace the subgraph. E.g. if there's an operator in subgraph that produces unbacked symints, the track_tensor_tree logic for that operator will take care of it.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] data-dependent op in body_fn"


The idea is the parent hop's fake tensor mode should ignore the newly allocated unbacked symints in subgraph because the bindings of unbacked symbols in the subgraph should already be done when we trace the subgraph. E.g. if there's an operator in subgraph that produces unbacked symints, the track_tensor_tree logic for that operator will take care of it.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] data-dependent op in body_fn"


The idea is the parent hop's fake tensor mode should ignore the newly allocated unbacked symints in subgraph because the bindings of unbacked symbols in the subgraph should already be done when we trace the subgraph. E.g. if there's an operator in subgraph that produces unbacked symints, the track_tensor_tree logic for that operator will take care of it.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/while_loop.py:
```diff
@@ -245,12 +245,64 @@ def _trace_while_loop(
 
 
+def check_outputs_carry_consistency(
+    outs: List[torch.Tensor], crys: List[torch.Tensor]
```
nit: crys -> carries

zou3519 commented on torch/_higher_order_ops/while_loop.py:
```diff
@@ -245,12 +245,64 @@ def _trace_while_loop(
+    if all_diffs_in_meta:
+        diff_str = "\n".join(all_diffs_in_meta)
+        raise RuntimeError(
+            f"Expected carried_inputs and body outputs return tensors with same metadata but find:\n{diff_str}"
```
nit: "but find" -> "but found"

ydwu4 committed:
Update on "[while_loop] data-dependent op in body_fn"


The idea is the parent hop's fake tensor mode should ignore the newly allocated unbacked symints in subgraph because the bindings of unbacked symbols in the subgraph should already be done when we trace the subgraph. E.g. if there's an operator in subgraph that produces unbacked symints, the track_tensor_tree logic for that operator will take care of it.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
